cpp/src/arrow/compute/exec.cc

```diff
@@ -156,15 +156,23 @@ Result<ExecBatch> ExecBatch::Make(std::vector<Datum> values) {
 
 Result<std::shared_ptr<RecordBatch>> ExecBatch::ToRecordBatch(
     std::shared_ptr<Schema> schema, MemoryPool* pool) const {
+  if (static_cast<size_t>(schema->num_fields()) > values.size()) {
+    return Status::Invalid("ExecBatch::ToTRecordBatch mismatching schema size");
+  }
   ArrayVector columns(schema->num_fields());
 
   for (size_t i = 0; i < columns.size(); ++i) {
     const Datum& value = values[i];
     if (value.is_array()) {
       columns[i] = value.make_array();
       continue;
+    } else if (value.is_scalar()) {
```

Comment:
Hmm..when would a ExecBatch contains scalar value and why do we convert that into array?

Reply:
This code was there to begin with, so I'll let others chime in. My guess is this is related to the `RecordBatch` API coming much earlier (and tied more closely to the Array API) than that of `ExecBatch`.

Reply:
RecordBatch requires Array columns. ExecBatch also supports for Scalars, which indicate a column with constant value. Such constant columns arise frequently (for example when the dataset has been partitioned on a column so a batch read from `cyl=4/dat.parquet` has a constant `cyl` column) and many kernels can take directly accept a scalar argument (for example the arithmetic kernels can add a scalar to an array), saving memory and processor overhead of broadcasting to Array.

Reply:
Right, it doesn't hurt to return a nice error here.